In [2]:
import pandas as pd
import numpy as np
import os

In [5]:
path=r'C:\Users\aditi\Desktop\Data Analytics_Career Foundry\Data Analytics_Immersion\Project 4\Instacart Basket Analysis'

In [9]:
df_ords=pd.read_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [14]:
# Mixed-Type Data
df_test=pd.DataFrame()

In [17]:
df_test['mix']=['a', 'b', 1, True]

In [20]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [23]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [26]:
df_prods=pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'products.csv'), index_col = False)

In [29]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [32]:
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [35]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [38]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [41]:
df_prods_clean.shape

(49677, 5)

In [44]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [47]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [51]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [54]:
df_prods_clean_no_dups.shape

(49672, 5)

In [58]:
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data', 'Prepared data', 'products_checked.csv'))

# Exercise 4.5

In [62]:
# Task 2

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

# 1. Count of rows in 'days_since_prior_order' is less than counts of rest of the columns, this means missing values.

# 2. Minimum value is 0 in 'days_since_prior_order' which could mean that orders were placed within the same day.

# 3. 'orders_day_of_week' has minimum value as 0 and maximum as 6 which means 0 is assigned to first day of the week eg. Monday or Sunday.

# 4. 'order_hour_of_day' is 0 which could either mean 00:00 midnight or this is a discrepancy which needs to be checked.

In [69]:
# Task 3

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

# There are no mixed data types in df_ords.

In [73]:
# Task 5
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

# 'days_since_prior_order' has 206209 missing values. This needs to be investigated as this column cannot be removed as it is an important variable.

In [81]:
# Task 6
df_na=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [83]:
df_na

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [85]:
df_ords_checking_order_number=(df_na['order_number']==1).all()

In [89]:
if df_ords_checking_order_number:
    print ("for every order having order_number 1, days_since_prior_order has value NaN")
else:
    print ("for every order having order_number 1, days_since_prior_order has a specific value")

for every order having order_number 1, days_since_prior_order has value NaN


# In case of order_number 1, days_since_prior_order is 'NaN' because these are the first orders and it makes sense that 'days_since_prior_order' is not applicable as these are the first order made by the customer and there aren't prior orders. So 'NaN' can be replaced by 0 in this case.

In [100]:
df_ords['days_since_prior_order'].fillna(0)

0           0.0
1          15.0
2          21.0
3          29.0
4          28.0
           ... 
3421078    29.0
3421079    30.0
3421080    18.0
3421081     7.0
3421082    30.0
Name: days_since_prior_order, Length: 3421083, dtype: float64

In [108]:
df_filled_na=df_ords[df_ords['order_number']==1]

In [110]:
df_filled_na

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     0.0  
11                      11                     0.0  
26                      14                     0.0  
39                      11                     0.0  
45                      12                     0.0  
...                    ...                     ...  
3420930                 12                     0.0  
3420934                 18                     0.0  
3421002                 19                     0.0  
3421019                 15                     0.0  
3421069                 11                     0.0  

[206209 rows x 7 columns]

In [112]:
df_ords.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

In [115]:
# Task 7
df_ords_dup = df_ords[df_ords.duplicated()]

In [117]:
df_ords_dup

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

# There are no duplicate rows in Orders_wrangled dataset.

In [124]:
df_ords.to_csv(os.path.join(path, 'Data', 'Prepared data', 'orders_cleaned.csv'))

In [126]:
df_ords=pd.read_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_cleaned.csv'), index_col = False)

In [128]:
df_ords.isnull().sum()

Unnamed: 0.1              0
Unnamed: 0                0
order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64